File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    bool backtrack(vector<int>& nums, int pos, vector<int>& memo) {
        if (pos == nums.size() - 1) return true;

        if (memo[pos] != -1) return memo[pos] == 1 ? true : false;

        for (int i = 1; i <= nums[pos]; i++) {
            if (backtrack(nums, pos + i, memo)) {
                memo[pos] = 1;
                return true;
            }
        }
        memo[pos] = 0;
        return false;
    }

    bool canJump(vector<int>& nums) {
        vector<int> memo(nums.size(), -1);

        return backtrack(nums, 0, memo);
    }
};

File: 0055-jump-game/0055-jump-game.py
from collections import deque

# class Solution:
#     def canJump(self, nums: List[int]) -> bool:
#         n = len(nums)
#         farthest_loc = 0
#         curr_loc = 0

#         while curr_loc <= farthest_loc:
#             farthest_loc = max(farthest_loc, curr_loc + nums[curr_loc])
#             if farthest_loc >= n - 1:
#                 return True
#             curr_loc += 1
    
#         return False

"""
test: nums = [2,3,1,1,4]
trace:
fl = 0, cl = 0
loop 0 <= 0
    fl = 2
    cl = 1
loop 1 <= 2
    fl = 4
    < returns True
"""

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        farthest_loc = 0

        for curr_loc in range(n):
            if farthest_loc < curr_loc:
                break
            
            farthest_loc = max(farthest_loc, curr_loc + nums[curr_loc])
            if farthest_loc >= n - 1:
                return True

        return False